<launch>
    <arg name="vehicle_name" default="bluerov" />
    <arg name="use_front_camera" default="false" />
    <arg name="use_vertical_camera" default="false" />
    <!-- known tags for AprilTag algorithm. Tags not specified in this file are ignored. Also
    specifies tags' size required for pose reconstruction. -->
    <arg name="tag_file_front_camera" default="$(find fav_sim)/config/tags_front_camera.yaml" />
    <arg name="tag_file_vertical_camera" default="$(find fav_sim)/config/tags_vertical_camera.yaml" />

    <!-- position -->
    <arg name="x" default="0.5" />
    <arg name="y" default="1.0" />
    <arg name="z" default="-0.5" />
    <!-- roll, pitch, yaw -->
    <arg name="R" default="0.0" />
    <arg name="P" default="0.0" />
    <arg name="Y" default="1.57" />
    <group ns="$(arg vehicle_name)">
        <param name="robot_description"
            command="$(find xacro)/xacro $(find fav_sim)/models/bluerov/urdf/bluerov.xacro use_front_camera:=$(arg use_front_camera) use_vertical_camera:=$(arg use_vertical_camera)" />
        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model"
            args="-param robot_description -urdf -model $(arg vehicle_name) -x $(arg x) -y $(arg y) -z $(arg z) -R $(arg R) -P $(arg P) -Y $(arg Y)" />
        <node name="esc_commander" pkg="fav_sim" type="esc_commander" output="screen" />
        <node name="mixer" pkg="fav_sim" type="simple_mixer" output="screen">
            <rosparam command="load" file="$(find fav_sim)/config/mixer_default.yaml" />
        </node>
        <node name="virtual_vision" pkg="fav_sim" type="virtual_vision_pose.py" output="screen">
            <param name="stddev_x" value="0.05" />
            <param name="stddev_y" value="0.05" />
            <param name="stddev_z" value="0.01" />
        </node>

        <group if="$(arg use_front_camera)" ns="front_camera">
            <node pkg="apriltag_ros"
                type="apriltag_ros_continuous_node"
                name="apriltag_node"
                clear_params="true"
                output="screen">
                <rosparam command="load" file="$(find apriltag_ros)/config/settings.yaml" />
                <rosparam command="load" file="$(arg tag_file_front_camera)" />
                <remap from="image_rect" to="image_raw" />
                <param name="camera_frame"
                    value="$(arg vehicle_name)/front_camera/camera_frame" />
                <param name="publish_tag_detections_image" value="true" />
            </node>
        </group>

        <group if="$(arg use_vertical_camera)" ns="vertical_camera">
            <node pkg="apriltag_ros"
                type="apriltag_ros_continuous_node"
                name="apriltag_node"
                clear_params="true"
                output="screen">
                <rosparam command="load" file="$(find apriltag_ros)/config/settings.yaml" />
                <rosparam command="load" file="$(arg tag_file_vertical_camera)" />
                <remap from="image_rect" to="image_raw" />
                <param name="camera_frame"
                    value="$(arg vehicle_name)/vertical_camera/camera_frame" />
                <param name="publish_tag_detections_image" value="true" />
            </node>
        </group>

    </group>
</launch>
